File: java/structural/BridgeDemo.java
//author sbkun
package structural;

abstract class Shape {
    Shape(Color color) {
        this.color = color;
    }

    abstract void draw();

    protected Color color;
}

abstract class Color {
    abstract String dye();
}

class Red extends Color {
    @Override
    String dye() {
        return "Red";
    }
}

class Blue extends Color{
    @Override
    String dye() {
        return "Blue";
    }
}

class Circle extends Shape {
    Circle(Color color) {
        super(color);
    }

    @Override
    public void draw() {
        System.out.println("draw Circle and dye it " + color.dye());
    }
}

class Rectangle extends Shape {
    Rectangle(Color color) {
        super(color);
    }

    @Override
    public void draw() {
        System.out.println("draw Rectangele and dye it " + color.dye());
    }
}

public class BridgeDemo {
    public static void main(String args[]) {
        Shape circle = new Circle(new Blue());
        Shape rectangle = new Rectangle(new Red());
        circle.draw();
        rectangle.draw();
    }
}

File: cpp/structural/BridgeDemo.cpp
//author sbkun
#include <iostream>
#include <stdexcept>
#include <memory>
#include <string>

using namespace std;

enum class ColorType {
	Red,
	Blue
};

class Color {
	public:
		virtual ~Color() = default;
		virtual string dye() = 0;	
};

class Red: public Color {
	public:
		string dye() override {
			return "Red";
		}
};

class Blue: public Color {
	public:
		string dye() override {
			return "Blue";
		}
};

class Shape {
	public:
		Shape(ColorType color) {
			switch(color) {
				case ColorType::Red:
					color_.reset(new Red());
					break;
				case ColorType::Blue:
					color_.reset(new Blue());
					break;
				default:
					color_.reset(new Red());
			}
		}

		virtual ~Shape() = default;
		virtual void draw() = 0;

	protected:
		unique_ptr<Color>	color_;
};

class Circle: public Shape {
	public:
		using Shape::Shape; //使用父类构造函数
		void draw() override {
			cout<<"draw "<<"Circle and dye it "<<color_->dye()<<endl;
		}	
};

class Rectangle: public Shape {
	public:
		using Shape::Shape; //使用父类构造函数
		void draw() override {
			cout<<"draw "<<"Rectangle and dye it "<<color_->dye()<<endl;
		}	
};

int main(int argc, char* argv[]) {
	unique_ptr<Shape> circle(new Circle(ColorType::Red));
	unique_ptr<Shape> rectangle(new Rectangle(ColorType::Blue));
	circle->draw();
	rectangle->draw();
}
